Don't allow strncmp'ing arbitrary vectors.

maleadt opened the pull request:
The current `startswith` can be called with arbitrary `Vector{UInt8}` inputs, an undocumented and unused use case as far as I can see. This is implemented in terms of `strncmp`, but according to the [POSIX spec](http://pubs.opengroup.org/onlinepubs/009695399/functions/strncmp.html) `bytes that follow a null byte are not compared`.

For example, current master yields the following incorrect result:

``` julia
julia> startswith(bytestring("foo\0bar"), bytestring("foo\0NO"))
true
```

I've opted to remove the `Vector{UInt8}` definition, which seems unused anyway, but I'm unsure about the remaining `ByteString` implementation. I've currently implemented it in terms of `ccall(:strncmp, ..., (Cstring, Cstring))` which disallows NULL characters altogether:

``` julia
> startswith(bytestring("foo\0bar"), bytestring("foo\0NO"))
ERROR: ArgumentError: embedded NUL chars are not allowed in C strings: "foo\0bar"
```

Maybe this should only be the case for `startswith(Cstring)`, with `startswith(ByteString)` resolving to the in-Julia version `startswith(AbstractString)`? Or just replace `strncmp` by `memcmp`?

maleadt committed:
Replace strncmp with memcmp when comparing strings with startswith().

vtjnash commented:
It would be better to switch to memcmp. Throwing an error for embedded null characters doesn't make much sense (and is requiring an extra pass over the string anyways to check for it).

vtjnash commented on base/strings/util.jl:
```diff
@@ -30,9 +30,8 @@ function endswith(a::AbstractString, b::AbstractString)
-startswith(a::Vector{UInt8}, b::Vector{UInt8}) =
-    (length(a) >= length(b) && ccall(:strncmp, Int32, (Ptr{UInt8}, Ptr{UInt8}, UInt), a, b, length(b)) == 0)
+startswith(a::ByteString, b::ByteString) =
+    (length(a) >= length(b) && ccall(:strncmp, Int32, (Cstring, Cstring, UInt), a, b, length(b)) == 0)
```
`sizeof(b) == length(b.data) != length(b)`

maleadt commented:
Ah yeah, did not think about that. I've reverted the old `ByteString` -> `Vector{UInt8}` indirection, just swapping `strncmp` with `memcmp`. I've also added a test.

vtjnash commented:
lgtm. thanks for finding and fixing this.

stevengj commented:
Maybe try a forced push?  The AppVeyor failure may be unrelated.
